Sources/SourceKitLSP/Swift/DiagnosticReportManager.swift

```diff
@@ -0,0 +1,183 @@
+  private nonisolated var requests: sourcekitd_requests { return sourcekitd.requests }
+
+  /// The cache that stores reportTasks for snapshot id and buildSettings
+  ///
+  /// Conceptually, this is a dictionary. To prevent excessive memory usage we
+  /// only keep `cacheSize` entries within the array. Older entries are at the
+  /// end of the list, newer entries at the front.
+  private var reportTaskCache:
+    [(
+      snapshotID: DocumentSnapshot.ID,
+      buildSettings: SwiftCompileCommand?,
+      reportTask: ReportTask
+    )] = []
+
+  /// The number of reportTasks to keep
+  ///
+  /// - Note: This has been chosen without scientific measurements.
+  private let cacheSize = 5
+
+  init(
+    sourcekitd: SourceKitD,
+    syntaxTreeManager: SyntaxTreeManager,
+    documentManager: DocumentManager,
+    clientHasDiagnosticsCodeDescriptionSupport: Bool
+  ) {
+    self.sourcekitd = sourcekitd
+    self.syntaxTreeManager = syntaxTreeManager
+    self.documentManager = documentManager
+    self.clientHasDiagnosticsCodeDescriptionSupport = clientHasDiagnosticsCodeDescriptionSupport
+  }
+
+  func diagnosticReport(
+    for snapshot: DocumentSnapshot,
+    buildSettings: SwiftCompileCommand?
+  ) async throws -> RelatedFullDocumentDiagnosticReport {
+    if let reportTask = reportTask(for: snapshot.id, buildSettings: buildSettings) {
+      return try await reportTask.value
+    }
+    let reportTask: Task<RelatedFullDocumentDiagnosticReport, Error>
+    if let buildSettings, !buildSettings.isFallback {
+      reportTask = Task {
+        return try await requestReport(with: snapshot, compilerArgs: buildSettings.compilerArgs)
+      }
+    } else {
+      logger.log(
+        "Producing syntactic diagnostics from the built-in swift-syntax because we \(buildSettings != nil ? "have fallback build settings" : "don't have build settings", privacy: .public))"
+      )
+      // If we don't have build settings or we only have fallback build settings,
+      // sourcekitd won't be able to give us accurate semantic diagnostics.
+      // Fall back to providing syntactic diagnostics from the built-in
+      // swift-syntax. That's the best we can do for now.
+      reportTask = Task {
+        return try await requestFallbackReport(with: snapshot)
+      }
+    }
+    setReportTask(for: snapshot.id, buildSettings: buildSettings, reportTask: reportTask)
+    return try await reportTask.value
+  }
+
+  func removeItemsFromCache(with uri: DocumentURI) async {
+    for item in reportTaskCache {
+      if item.snapshotID.uri == uri {
+        item.reportTask.cancel()
+      }
+    }
+    reportTaskCache.removeAll(where: { $0.snapshotID.uri == uri })
+  }
+
+  private func requestReport(
+    with snapshot: DocumentSnapshot,
+    compilerArgs: [String]
+  ) async throws -> LanguageServerProtocol.RelatedFullDocumentDiagnosticReport {
+    try Task.checkCancellation()
+
+    let keys = self.keys
```

Comment:
I don’t think this is necessary. `keys` should already refer to `self.keys` everywhere below.

Reply:
i checked the var keys the implementations is 
```
private nonisolated var keys: sourcekitd_keys { return sourcekitd.keys }
```
i think the purpose is to ensure the value of keys won't change in func requestReport

the same pattern is happening in other files. 
e.g. SwiftLanguageServer.swift, SemanticRefactoring.swift, etc.

Reply:
Ah, OK. Fair enough.